pkg/deploy/registry/instantiate/rest.go

```diff
@@ -0,0 +1,180 @@
+	req, ok := obj.(*deployapi.DeploymentRequest)
+	if !ok {
+		return nil, false, errors.NewBadRequest(fmt.Sprintf("wrong object passed to DeploymentRequest update: %#v", obj))
+	}
+
+	if errs := validation.ValidateDeploymentRequest(req); len(errs) > 0 {
+		return nil, false, errors.NewInvalid(deployapi.Kind("DeploymentRequest"), req.Name, errs)
+	}
+
+	configObj, err := r.store.Get(ctx, req.Name)
+	if err != nil {
+		return nil, false, err
+	}
+	config := configObj.(*deployapi.DeploymentConfig)
+
+	// Redeploy what we have.
+	if !req.Latest {
+		if req.Manual {
+			config.Status.Details = new(deployapi.DeploymentDetails)
+			config.Status.Details.Causes = []deployapi.DeploymentCause{{Type: deployapi.DeploymentTriggerManual}}
+		}
+		config.Status.LatestVersion++
+		return r.dryRunOrUpdate(ctx, config, req.DryRun)
+	}
+
+	// We need to process the deployment config before we can determine if it is possible to trigger
+	// a deployment. This is true for the generic trigger controller. Manual deployments should
+	// kick off a deployment unconditionally.
+	if err := r.process(config, req.Manual); err != nil && !req.Manual {
+		return nil, false, errors.NewInternalError(err)
+	}
+
+	decoded, err := r.decodeFromLatest(config)
+	if err != nil {
+		return nil, false, err
+	}
+
+	var (
+		canTrigger bool
+		causes     []deployapi.DeploymentCause
+	)
+
+	canTrigger, causes = deployutil.CanTrigger(config, decoded)
+	if req.Manual {
+		causes = []deployapi.DeploymentCause{{Type: deployapi.DeploymentTriggerManual}}
+	}
+
+	if canTrigger || req.Manual {
+		config.Status.Details = new(deployapi.DeploymentDetails)
+		config.Status.Details.Causes = causes
+		config.Status.LatestVersion++
+	}
+	return r.dryRunOrUpdate(ctx, config, req.DryRun)
+}
+
+func (r *REST) dryRunOrUpdate(ctx kapi.Context, config *deployapi.DeploymentConfig, dryRun bool) (runtime.Object, bool, error) {
+	if dryRun {
+		return config, false, nil
+	}
+	return r.store.Update(ctx, config)
+}
+
+// process will go over all deployment triggers that require processing and update the
+// deployment config accordingly. This is the work that the image change controller
+// was doing up to this point.
+func (r *REST) process(config *deployapi.DeploymentConfig, manual bool) error {
+	errs := []error{}
+	// Process any image change triggers.
+	for _, trigger := range config.Spec.Triggers {
+		if trigger.Type != deployapi.DeploymentTriggerOnImageChange {
+			continue
+		}
+
+		params := trigger.ImageChangeParams
+
+		// Manual deployments should always try to resolve the images in the template.
+		// On the other hand, paused deployments or non-automatic triggers shouldn't.
+		if !manual && config.Spec.Paused || !params.Automatic {
+			continue
+		}
+
+		// Tag references are already validated
+		name, tag, _ := imageapi.SplitImageStreamTag(params.From.Name)
+		istag, err := r.istn.ImageStreamTags(params.From.Namespace).Get(name, tag)
```

Comment:
Primary reason is efficiency - if you need two tags from the same repo
(two different triggers) then IST is 4 times the number of calls (each
IST is two loads from etcd).

For infrastructure code if you are looking up tags you should generally call IS